elastic-search-support/src/main/java/uk/gov/hmcts/ccd/definition/store/elastic/ElasticDefinitionImportListener.java

```diff
@@ -41,42 +44,116 @@ public ElasticDefinitionImportListener(CcdElasticSearchProperties config, CaseMa
+
+                    //initiate reindexing
+                    CompletableFuture<String> taskId = handleReindexing(elasticClient, baseIndexName, caseTypeName,
+                        incrementedCaseTypeName, deleteOldIndex);
+                    event.setTaskId(taskId.get());
+                } else {
+                    caseMapping = mappingGenerator.generateMapping(caseType);
+                    log.debug("case mapping: {}", caseMapping);
+                    elasticClient.upsertMapping(baseIndexName, caseMapping);
+                }
             }
         } catch (ElasticsearchStatusException exc) {
             logMapping(caseMapping);
             throw elasticsearchErrorHandler.createException(exc, currentCaseType);
+        } catch (InterruptedException ie) {
+            Thread.currentThread().interrupt();
+            logMapping(caseMapping);
         } catch (Exception exc) {
             logMapping(caseMapping);
             throw new ElasticSearchInitialisationException(exc);
-        } finally {
-            if (elasticClient != null) {
-                elasticClient.close();
-            }
         }
     }
 
+    private CompletableFuture<String> handleReindexing(HighLevelCCDElasticClient elasticClient, String baseIndexName,
+                                                       String caseTypeName, String incrementedCaseTypeName,
+                                                       boolean deleteOldIndex) {
+        //initiate async elasticsearch reindexing request
+        CompletableFuture<String> reindexFuture = elasticClient.reindexData(caseTypeName, incrementedCaseTypeName)
+            .exceptionally(ex -> {
+                try {
+                    //if failed delete new index, set old index writable
+                    HighLevelCCDElasticClient asyncElasticClient = clientFactory.getObject();
+                    log.info("reindexing failed, error: {}", ex.getMessage());
+                    asyncElasticClient.removeIndex(incrementedCaseTypeName);
+                    log.info("{} deleted", incrementedCaseTypeName);
+                    asyncElasticClient.setIndexReadOnly(caseTypeName, false);
+                    log.info("{} set to writable", caseTypeName);
+                } catch (IOException e) {
+                    log.error("failed to clean up after reindexing failure", e);
+                    throw new CompletionException(e);
+                }
+                throw new CompletionException(ex);
+            });
+
+        //if success set writable and update alias to new index
+        return reindexFuture.thenApply(taskId -> {
+            try {
+                HighLevelCCDElasticClient asyncElasticClient = clientFactory.getObject();
+                log.info("updating alias from {} to {}", caseTypeName, incrementedCaseTypeName);
+                asyncElasticClient.setIndexReadOnly(baseIndexName, false);
```

Comment:
baseIndexName has to be in readonly state?
                asyncElasticClient.setIndexReadOnly(baseIndexName, true);